tracecat/auth/saml.py

```diff
@@ -0,0 +1,289 @@
+
+    # Create IDPSSODescriptor element
+    idp_sso_descriptor = ET.SubElement(
+        root,
+        "IDPSSODescriptor",
+        {"protocolSupportEnumeration": "urn:oasis:names:tc:SAML:2.0:protocol"},
+    )
+
+    # Add KeyDescriptor
+    key_descriptor = ET.SubElement(idp_sso_descriptor, "KeyDescriptor", use="signing")
+    key_info = ET.SubElement(key_descriptor, "ds:KeyInfo")
+    x509_data = ET.SubElement(key_info, "ds:X509Data")
+    x509_certificate = ET.SubElement(x509_data, "ds:X509Certificate")
+    x509_certificate.text = SAML_IDP_CERTIFICATE
+
+    # Add NameIDFormat
+    name_id_format = ET.SubElement(idp_sso_descriptor, "NameIDFormat")
+    name_id_format.text = "urn:oasis:names:tc:SAML:2.0:nameid-format:persistent"
+
+    # Add SingleSignOnService
+    ET.SubElement(
+        idp_sso_descriptor,
+        "SingleSignOnService",
+        {"Binding": BINDING_HTTP_REDIRECT, "Location": SAML_IDP_REDIRECT_URL},
+    )
+
+    # Create a temporary file
+    with tempfile.NamedTemporaryFile(
+        mode="w+", suffix=".xml", delete=False
+    ) as tmp_file:
+        # Write the XML to the temporary file
+        tree = ET.ElementTree(root)
+        tree.write(tmp_file, encoding="unicode", xml_declaration=True)
+        tmp_file.flush()
+        tmp_file_path = tmp_file.name
+
+    try:
+        yield tmp_file_path
+    finally:
+        # Clean up the temporary file
+        os.unlink(tmp_file_path)
+
+
+def create_saml_client() -> Saml2Client:
+    saml_settings = {
+        "strict": True,
+        # The global unique identifier for this service provider
+        "entityid": TRACECAT__PUBLIC_API_URL,
+        "xmlsec_binary": XMLSEC_BINARY_PATH,
+        # Service provider settings
+        "service": {
+            "sp": {
+                "name": "tracecat_saml_sp",
+                "description": "Tracecat SAML SSO Service Provider",
+                "endpoints": {
+                    "assertion_consumer_service": [
+                        (SAML_SP_ACS_URL, BINDING_HTTP_POST),
+                    ],
+                },
+                # Security settings
+                "allow_unsolicited": True,  # If true, it allows the IdP to initiate the authentication (not yet supported)
+                "authn_requests_signed": False,  # Don't need to sign authn requests because we don't control the IdP
+                "want_assertions_signed": True,  # We require the IdP to sign the assertions
+                "want_response_signed": False,
+            },
+        },
+    }
+
+    if SAML_IDP_METADATA_URL is None:
+        with generate_saml_metadata_file() as tmp_metadata_path:
+            # Add the local metadata file to the settings
+            saml_settings["metadata"] = {"local": [tmp_metadata_path]}
+            config = Saml2Config()
+            config.load(saml_settings)
+    else:
+        # Save the cert to a temporary file
+        with tempfile.NamedTemporaryFile(
+            mode="w+", suffix=".crt", delete=False
```

Comment:
wondering if its ok to skip cleaning up these certs? a nice pattern could be we turn this function into a fastapi dependency and `yield` inside this context manager - then when the fastapi router is done handling this route, it will cleanup the file